### PR TITLE
Lo6165 price bump cap

### PR DESCRIPTION
**Pull request by dai1975**
- fix price bump bug to set opts.Nonce before GasFeeCalculator#Apply
 - check latest(replacing) transaction's gas fee is already high enough
 - make IETHClient interface and use it in gas.go and txpool.go for testing

### CODE REVIEW

File: pkg/client/client.go [hunk 1]
```diff
@@ -16,6 +16,41 @@ import (
...
+	ChainID             *hexutil.Big      `json:"chainId,omitempty"`
+	BlobVersionedHashes []common.Hash     `json:"blobVersionedHashes,omitempty"`
+	V                   *hexutil.Big      `json:"v"`
+	R                   *hexutil.Big      `json:"r"`
+	S                   *hexutil.Big      `json:"s"`
+	YParity             *hexutil.Uint64   `json:"yParity,omitempty"`
+}
+
+// wrapping interface of ethclient.Client struct
+type IETHClient interface {
```
Comment (dai1975) [inline pkg/client/client.go:46]: Define interface for testing.
Currently functions only used in gas.go and txpool.go are declared.

File: pkg/client/txpool/txpool.go [hunk 2]
```diff
@@ -82,30 +45,29 @@ func inclByPercent(n *big.Int, percent uint64) {
...
 	}
 	if targetTx == nil {
-		return common.Big0, common.Big0, nil
+		return nil, common.Big0, common.Big0, nil
 	}
-
-	gasFeeCap := targetTx.GasFeeCap.ToInt()
-	gasTipCap := targetTx.GasTipCap.ToInt()
+	gasFeeCap := new(big.Int).Set(targetTx.GasFeeCap.ToInt())
+	gasTipCap := new(big.Int).Set(targetTx.GasTipCap.ToInt())
```
Comment (dai1975) [inline pkg/client/txpool/txpool.go:67]: hexuti.Big.ToInt() returns same pointer of receiver instance and following inclByPercent() overwrite targetTx. This make problems when targetTx object is stored and reused such as testing mock.

**Comment by siburu** [comment]
@dai1975 You have replaced the `ETHClient` struct with the `IETHClient` interface to mock it for testing. They are all defined in the `pkg/client` package.
However, actual tests that want to mock the ethereum client are not in `pkg/client` but in `pkg/relay/ethereum` (`gas_test.go`).
Therefore, I think, you didn't need to modify the `pkg/client` package and should only have modified the type of the `client` argument of `NewGasFeeCalculator`.

**Comment by siburu** [comment]
@dai1975 To do so, you may need to move the functions defined in the `pkg/client/txpool` package to the `pkg/client` and convert them into methods of the `ETHClient` struct.

File: pkg/relay/ethereum/client.go [hunk 3]
```diff
@@ -38,5 +40,34 @@ func (chain *Chain) TxOpts(ctx context.Context, useLatestNonce bool) (*bind.Tran
 		}
 	}
 
+	if err := NewGasFeeCalculator(chain.client, &chain.config).Apply(ctx, txOpts); err != nil {
+		return nil, err
+	}
+
```
Comment (dai1975) [inline pkg/relay/ethereum/client.go:46]: statement move to fix bug

**Comment by dai1975** [comment]
@siburu moved interface definition to ethereum package.

File: pkg/relay/ethereum/client.go [hunk 4]
```diff
@@ -38,5 +40,38 @@ func (chain *Chain) TxOpts(ctx context.Context, useLatestNonce bool) (*bind.Tran
...
+// implements IETHChainClient
+func (cl *ChainClient) SuggestGasPrice(ctx context.Context) (*big.Int, error) {
+	return cl.ETHClient.Client.SuggestGasPrice(ctx)
+}
+func (cl *ChainClient) HeaderByNumber(ctx context.Context, number *big.Int) (*gethtypes.Header, error) {
+	return cl.ETHClient.Client.HeaderByNumber(ctx, number)
+}
+func (cl *ChainClient) FeeHistory(ctx context.Context, blockCount uint64, lastBlock *big.Int, rewardPercentiles []float64) (*ethereum.FeeHistory, error) {
+	return cl.ETHClient.Client.FeeHistory(ctx, blockCount, lastBlock, rewardPercentiles)
+}
```
[the remark this answers is not in the document]
Comment (dai1975) [reply pkg/relay/ethereum/client.go]: Sorry, these codes are remains of try of struct version. Removed.

File: pkg/relay/ethereum/client.go [hunk 5]
```diff
@@ -38,5 +40,38 @@ func (chain *Chain) TxOpts(ctx context.Context, useLatestNonce bool) (*bind.Tran
...
 	return txOpts, nil
 }
+
+// wrapping interface of client.ETHClient struct
+type IChainClient interface {
+	SuggestGasPrice(ctx context.Context) (*big.Int, error)
+	HeaderByNumber(ctx context.Context, number *big.Int) (*gethtypes.Header, error);
+	FeeHistory(ctx context.Context, blockCount uint64, lastBlock *big.Int, rewardPercentiles []float64) (*ethereum.FeeHistory, error)
+	GetMinimumRequiredFee(ctx context.Context, address common.Address, nonce uint64, priceBump uint64) (*txpool.RPCTransaction, *big.Int, *big.Int, error);
+}
```
[the remark this answers is not in the document]
Comment (dai1975) [reply pkg/relay/ethereum/client.go]: I did not know it. fixed.

File: pkg/relay/ethereum/gas.go [hunk 6]
```diff
@@ -61,6 +66,9 @@ func (m *GasFeeCalculator) Apply(ctx context.Context, txOpts *bind.TransactOpts)
 		// GasFeeCap = min(LimitFeePerGas, GasTipCap + BaseFee * BaseFeeRate)
 		m.config.DynamicTxGasConfig.BaseFeeRate.Mul(gasFeeCap)
 		gasFeeCap.Add(gasFeeCap, gasTipCap)
+		if oldTx != nil && oldTx.GasFeeCap != nil && oldTx.GasFeeCap.ToInt().Cmp(gasFeeCap) > 0 {
+			return fmt.Errorf("old tx's gasFeeCap(%v) is higher than suggestion(%v)", oldTx.GasFeeCap.ToInt(), gasFeeCap)
+		}
```
Comment (siburu) [inline pkg/relay/ethereum/gas.go:71]: @dai1975 I think that we should give up tx replacement only if both the tip and fee caps of the pending tx are higher than the new tx.

Comment (dai1975) [reply pkg/relay/ethereum/gas.go]: Surely. fixed.

Comment (dai1975) [reply pkg/relay/ethereum/gas.go]: add error case when new caps are equals to old caps(Cmp returns zero)

File: pkg/relay/ethereum/gas.go [hunk 7]
```diff
@@ -52,15 +54,24 @@ func (m *GasFeeCalculator) Apply(ctx context.Context, txOpts *bind.TransactOpts)
 		}
 		// GasTipCap = min(LimitPriorityFeePerGas, simulated_eth_maxPriorityFeePerGas * PriorityFeeRate)
 		m.config.DynamicTxGasConfig.PriorityFeeRate.Mul(gasTipCap)
+
+		// GasFeeCap = min(LimitFeePerGas, GasTipCap + BaseFee * BaseFeeRate)
+		m.config.DynamicTxGasConfig.BaseFeeRate.Mul(gasFeeCap)
+		gasFeeCap.Add(gasFeeCap, gasTipCap)
+
+		if oldTx != nil && oldTx.GasFeeCap != nil && oldTx.GasTipCap != nil {
+			if oldTx.GasFeeCap.ToInt().Cmp(gasFeeCap) > 0 && oldTx.GasTipCap.ToInt().Cmp(gasTipCap) > 0 {
```
[the remark this answers is not in the document]
Comment (dai1975) [reply pkg/relay/ethereum/gas.go]: fixed.

File: pkg/relay/ethereum/gas.go [hunk 8]
```diff
@@ -52,15 +54,24 @@ func (m *GasFeeCalculator) Apply(ctx context.Context, txOpts *bind.TransactOpts)
...
 		// GasTipCap = min(LimitPriorityFeePerGas, simulated_eth_maxPriorityFeePerGas * PriorityFeeRate)
 		m.config.DynamicTxGasConfig.PriorityFeeRate.Mul(gasTipCap)
+
+		// GasFeeCap = min(LimitFeePerGas, GasTipCap + BaseFee * BaseFeeRate)
+		m.config.DynamicTxGasConfig.BaseFeeRate.Mul(gasFeeCap)
+		gasFeeCap.Add(gasFeeCap, gasTipCap)
+
+		if oldTx != nil && oldTx.GasFeeCap != nil && oldTx.GasTipCap != nil {
+			if oldTx.GasFeeCap.ToInt().Cmp(gasFeeCap) >= 0 && oldTx.GasTipCap.ToInt().Cmp(gasTipCap) >= 0 {
+				return fmt.Errorf("old tx's gasFeeCap(%v) and gasTipCap(%v) are higher than suggestion(%v, %v)", oldTx.GasFeeCap.ToInt(), oldTx.GasTipCap.ToInt(), gasFeeCap, gasTipCap)
```
Comment (siburu) [inline pkg/relay/ethereum/gas.go:64]: Please fix the error message by replacing `higher than` with `greater than or equal to`.

File: pkg/relay/ethereum/gas_test.go [hunk 9]
```diff
@@ -129,3 +134,171 @@ func Test_getFeeInfo(t *testing.T) {
...
+			t.Fatal(err)
+		}
+	}
+
+	// test that old both tx's gasFeeCap and gasTipCap exceed suggestion
+	{
+		cli.MockPendingTransaction.GasTipCap = (*hexutil.Big)(big.NewInt(100))
+		cli.MockPendingTransaction.GasFeeCap = (*hexutil.Big)(big.NewInt(300))
+		cli.MockHistoryGasTipCap.SetUint64(99) // lower than 100
+		cli.MockHistoryGasFeeCap.SetUint64(299 - 99) //lower than 300
```
Comment (siburu) [inline pkg/relay/ethereum/gas_test.go:298]: You should replace `99` and `299 - 99` with `100` and `300 - 100` to test the corner case.

**Review by siburu — COMMENTED** [review]
@dai1975 Sorry but I have specified some to fix. Please check them :bow:

File: pkg/relay/ethereum/gas.go (hunk 8 above)
Comment (dai1975) [reply pkg/relay/ethereum/gas.go]: fixed

File: pkg/relay/ethereum/gas_test.go (hunk 9 above)
Comment (dai1975) [reply pkg/relay/ethereum/gas_test.go]: fixed
